src/data_processor.py

```diff
@@ -1,6 +1,160 @@
-"""Process data retrieved by data_miner.py."""
-# import data_collection
+"""A file for calculating how well a team works together using the interquartile range.
 
+and calculate the individual contribution as a percentage of overall contribution
+"""
 
-def process_data():
-    """Process the calculated data."""
+# This import fixes a linting error with old division.
+from __future__ import division
+from collections import defaultdict
+import numpy as np
+import data_collection
+
+
+def iterate_nested_dictionary(dictionary):
+    """Iterate through the nested dictionary and create new dictionary."""
+    # iterate through nested dictionary and add to new dictionary
+    category_scores = {}
+    amount = 0
+    for user in dictionary.keys():
+        for metric in dictionary[user]:
+            # disregard any instances of strings in dictionary
+            if not isinstance(dictionary[user][metric], str):
+                # if list, add the length of the list to new dictionary
+                if isinstance(dictionary[user][metric], list):
+                    amount = len(dictionary[user][metric])
+                    # print(amount)
+                    if metric not in category_scores:
+                        category_scores[metric] = [amount]
+                    else:
+                        category_scores[metric].append(amount)
+                # if not list, add values to new dictionary
+                else:
+                    if metric not in category_scores:
+                        category_scores[metric] = [dictionary[user][metric]]
+                    else:
+                        category_scores[metric].append(dictionary[user][metric])
+
+    # return the new dictionary
+    return category_scores
+
+
+def calculate_iqr_score(data_list, below_weight, above_weight, within_weight):
+    """Calculate a team score with interquartile range."""
+    # pylint: disable=too-many-locals
+    below_amount = 0
+    above_amount = 0
+    within_amount = 0
+    dataset = data_list
+    # calculate the size of the dataset
+    size = len(dataset)
+    # sort the dataset in ascending order
+    sorted(dataset)
+    # determine the first and third quartiles
+    q1, q3 = np.percentile(dataset, [25, 75])
+    # calculate the interquartile range (q3-q1)
+    iqr = q3 - q1
+    # calculate the lower bound of the IQR
+    lower_bound = q1 - (1.5 * iqr)
+    # calculate the upper bound of the IQR
+    upper_bound = q3 + (1.5 * iqr)
+    # for any value above, below or within the IQR add to the total
+    for d in dataset:
+        if d < lower_bound:
+            below_amount = below_amount + 1
+        if d > upper_bound:
+            above_amount = above_amount + 1
+        if lower_bound <= d <= upper_bound:
+            within_amount = within_amount + 1
```

Comment:
From my understanding of IQR and the box and whisker plot, isn't it the case that 25% of the data points will be below the IQR, 50% within the IQR, and 25% above it. In that case, why do you need to count data points above, within, or below the IQR?

Reply:
IQR shows the spread of the data, and if there are any datapoints that don't fit into the trend of the rest of the data (which are called outliers). It is not the case that there is a specific amount of data below, above, and within, the iqr, which is why this program calculates the amount below, above, and within. We want as many points within the iqr as possible, as this means that everyone spread the work pretty evenly throughout the team.